Add contrib playbook to disable service firewall

oomichi opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Basically we need to make necessary TCP/UDP ports open.
However the necessary ports are so many, and sometimes it is difficult to figure out that is due to firewall issues or not if facing deployment issues.
To distinguish a root problem on such situation, this adds a contrib playbook to disable the service firewall for Kubespray development and test.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
The contrib playbook os-manage is added to disable service firewall for Kubespray development and test.
```

oomichi commented:
confirmed this option works on RHEL8 of Azure vm.

champtar commented on roles/bootstrap-os/tasks/bootstrap-centos.yml:
```diff
@@ -93,3 +93,17 @@
   become: true
+
+- name: Stop service firewalld, if started
+  service:
```
you can write this as 1 task no ? for sure with `systemd:` you can

oomichi replied:
Are you saying like
```
- name: Disable service firewalld
  systemd:
      name: firewalld
      state: stopped
      enabled: no
```

oomichi replied:
Thanks for your advice, let me try and test it.

oomichi replied:
confirmed that works, nice advice :)

champtar replied:
you are welcome

champtar reviewed (commented):
what do you think about having the default to false, but this value in the sample inventory to true ? that would not break existing cluster but remove one classic bug on many first deployments ? we might need to check if firewalld is installed in that case.

oomichi commented:
> 
> 
> what do you think about having the default to false, but this value in the sample inventory to true ? that would not break existing cluster but remove one classic bug on many first deployments ? we might need to check if firewalld is installed in that case.

Basically operators need to open necessary TCP/UDP ports on enabled firewall service, I feel.
So I hesitate to have `true` on the Kubespray repo.

champtar commented:
Most firewalld rule modifications will trigger a reload and break K8S rules when using iptables. On CentOS 8 firewalld is using nftables directly so I think it doesn't break K8S rules, but I moved to Calico rules (that have some other corner cases)

oomichi commented:
/cc @floryut @EppO

ant31 commented:
I would be against it, 
The problem is more a lack of documentation rather than a technical one.

Firewall configuration (and even more deactivation) has been out of scope. 
You can always run your own ansible-playbook to disable the firewall, and eventually, to contribute it to the directory https://github.com/kubernetes-sigs/kubespray/tree/master/contrib. But even there, I would rather see a playbook to configure firewalls with the necessary ports, which could be helpful for everyone to understand what's the minimum ports configuration.

oomichi commented:
> 
> 
> I would be against it,
> The problem is more a lack of documentation rather than a technical one.
> 
> Firewall configuration (and even more deactivation) has been out of scope.
> You can always run your own ansible-playbook to disable the firewall, and eventually, to contribute it to the directory https://github.com/kubernetes-sigs/kubespray/tree/master/contrib. But even there, I would rather see a playbook to configure firewalls with the necessary ports, which could be helpful for everyone to understand what's the minimum ports configuration.

I agree to configure firewalls with the necessary ports.
Then I wrote this PR message like
```
Basically we need to make necessary TCP/UDP ports open.
However the necessary ports are so many, and sometimes it is difficult to figure out that is due to firewall issues or not if facing deployment issues.
To distinguish a root problem on such situation, this adds an option to disable the service firewall for Kubespray development and test.
```
When I faced this related issue, it was hard to find that is due to firewall issue or not according to the error message.
Then I created this PR to distinguish a root problem easily by setting this feature disabled for future works which contain PRs to make some TCP/UDP ports open on Kubespray project.

ant31 commented:
> Then I created this PR to distinguish a root problem easily by setting this feature disabled for future works which contain PRs to make some TCP/UDP ports open on Kubespray project.

Yes, thank you for looking at this. In between, I don't think that this feature is needed in the main playbook as it's more a debug tool. 
You could propose a separate playbook: 
```
ansible-playbook contrib/manage-firewalld -e "{disable_firewalld: true}" -i inventory/cluster.yaml 
```

oomichi commented:
> 
> 
> > Then I created this PR to distinguish a root problem easily by setting this feature disabled for future works which contain PRs to make some TCP/UDP ports open on Kubespray project.
> 
> Yes, thank you for looking at this. In between, I don't think that this feature is needed in the main playbook as it's more a debug tool.
> You could propose a separate playbook:
> 
> ```
> ansible-playbook contrib/manage-firewalld -e "{disable_firewalld: true}" -i inventory/cluster.yaml 
> ```

That is a good idea :-)

oomichi commented:
To convert this to contrib

/hold

oomichi commented:
confirmed this contrib playbook works on both RHEL8 and Ubuntu 18.04:
```
$ ansible-playbook -e "{disable_service_firewall: true}" -i inventory/sample/hosts.yaml --become --become-user=root contr
ib/os-services/os-services.yml 

PLAY [all] *******************************************************************************************************************************************
Monday 05 April 2021  23:52:43 +0000 (0:00:00.065)       0:00:00.065 ********** 

TASK [prepare : Disable service firewalld (RedHat)] **************************************************************************************************
changed: [rhel8-1]
Monday 05 April 2021  23:52:45 +0000 (0:00:02.037)       0:00:02.102 ********** 

TASK [prepare : Disable service ufw (Debian)] ********************************************************************************************************
changed: [ubuntu-02]

PLAY RECAP *******************************************************************************************************************************************
rhel8-1                    : ok=1    changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
ubuntu-02                  : ok=1    changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   

Monday 05 April 2021  23:52:46 +0000 (0:00:00.960)       0:00:03.062 ********** 
=============================================================================== 
prepare : Disable service firewalld (RedHat) -------------------------------------------------------------------------------------------------- 2.04s
prepare : Disable service ufw (Debian) -------------------------------------------------------------------------------------------------------- 0.96s
$ sudo systemctl status firewalld
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
     Docs: man:firewalld(1)

Apr 05 23:40:16 localhost.localdomain systemd[1]: Starting firewalld - dynamic firewall daemon...
Apr 05 23:40:19 localhost.localdomain systemd[1]: Started firewalld - dynamic firewall daemon.
Apr 05 23:40:20 localhost.localdomain firewalld[959]: WARNING: AllowZoneDrifting is enabled. This is considered an insecure configuration option. It >
Apr 05 23:52:44 rhel8-1 systemd[1]: Stopping firewalld - dynamic firewall daemon...
Apr 05 23:52:45 rhel8-1 systemd[1]: Stopped firewalld - dynamic firewall daemon.
```

/hold cancel

champtar commented on contrib/os-services/roles/prepare/tasks/main.yml:
```diff
@@ -0,0 +1,18 @@
+    enabled: no
+  when:
+    - disable_service_firewall
+    - "ansible_os_family == 'Debian'"
```
Debian can have ufw or firewalld, so better list existing systemd services and disable firewalld or ufw if present

oomichi replied:
I see, let me add the check.

champtar commented:
Also stopping a firewall config daemon might require a restart for everything to fall back in place

oomichi commented:
> 
> 
> Also stopping a firewall config daemon might require a restart for everything to fall back in place

I don't face issues without restarting during this work.
I'd like to keep this without that if possible because I am testing this PR on the same machine.

champtar commented:
> > Also stopping a firewall config daemon might require a restart for everything to fall back in place
> 
> I don't face issues without restarting during this work.
> I'd like to keep this without that if possible because I am testing this PR on the same machine.

Sorry I meant 'if k8s is already running', but if you don't see issue that's good

liupeng0518 commented on contrib/os-services/roles/prepare/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+      "'ufw.service' in services"
+
+  when:
+  - disable_service_firewall
```
Should we add a default value os-services/roles/prepare/defaults/main.yml `disable_service_firewall: true/false`

oomichi replied:
That is a nice point, I will add it.

oomichi replied:
Done

ant31 commented:
look good, but why calling it os-services?  I would go with firewall or something very specific to the usecase

oomichi commented:
> 
> 
> look good, but why calling it os-services? I would go with firewall or something very specific to the usecase

I guess we will want to add the other services also like selinux, apparmor etc. in the future.

floryut reviewed (approved):
Contrib play so let's merge, this could always be improved afterward if need be.

champtar commented:
/lgtm
